libutils/passopenfile.c

```diff
@@ -0,0 +1,509 @@
+    char msg[PID_MSG_SIZE];
+    assert(sizeof(PID_FMT) - 3 + PRINTSIZE(unsigned long) <= PID_MSG_SIZE);
+    unsigned long pid = GetCurrentProcessId();
+    int len = snprintf(msg, sizeof(msg), PID_FMT, pid);
+    assert(len > 0 && len < PID_MSG_SIZE);
+    send(uds, msg, len + 1, 0);
+
+    bool ready;
+    if (!wait_for(uds, false, &ready))
+    {
+        Log(LOG_LEVEL_ERR,
+            "Can't receive descriptor (select: %s)",
+            GetErrorStr());
+        return -1;
+    }
+    else if (!ready)
+    {
+        Log(LOG_LEVEL_VERBOSE, "No descriptor received.");
+        return -1;
+    }
+    else
+    {
+        WSAPROTOCOL_INFO blob;
+
+        /* Receive blob and text over UDS: */
+        ssize_t got = recv(uds, (const void*)&blob, sizeof(blob), 0);
+        /* Casts needed because MinGW thinks recv(,const char*,,) :-( */
+        if (got == sizeof(blob))
+        {
+            descriptor = WSASocket(FROM_PROTOCOL_INFO,
+                                   FROM_PROTOCOL_INFO,
+                                   FROM_PROTOCOL_INFO,
+                                   /* Args after the blob are ignored. */
+                                   &blob, 0, WSA_FLAG_OVERLAPPED);
+            if (descriptor == SOCKET_ERROR)
+            {
+                return -1;
+            }
+            /* else we *must* either use or closesocket() this descriptor. */
+
+            if (text)
+            {
+                *text = NULL;
+            }
+            MSG_LEN_T size; /* Including space for final '\0' byte: */
+            got = recv(uds, (const void*)&size, sizeof(size), 0);
+            if (got == sizeof(size))
+            {
+                if (size)
+                {
+                    char *buffer = malloc(size);
+                    if (*text)
+                    {
+                        got = recv(uds, buffer, size, 0);
+                        if (got != size)
+                        {
+                            Log(LOG_LEVEL_ERR,
+                                "Failed to receive whole text accompanying descriptor");
+                        }
+                    }
+                    else
+                    {
+                        got = 0; /* != size */
+                        Log(LOG_LEVEL_ERR,
+                            "Failed to allocate buffer for text accompanying descriptor");
+                    }
+
+                    if (text && got == size)
+                    {
+                        *text = buffer;
+                    }
+                    else
+                    {
+                        free(buffer);
+                    }
+                }
+            }
+            else
+            {
+                Log(LOG_LEVEL_ERR,
+                    "Failed to read size of text accompanying descriptor");
+                size = 0;
+            }
+
+            if (text && size && !*text)
+            {
+                closesocket(descriptor);
+                return -1;
+            }
```

Comment:
The case you're thinking of is an empty string: size is 1, *text is a non-NULL pointer and **text is '\0'.  The size includes trailing '\0' so that we can distinguish empty string from NULL (size == 0).